### PR TITLE
fix(flags): Retry flag requests on transient network errors

### LINKED ISSUE #195: Feature flag evaluation request has no retry on transient network errors
The feature flag evaluation request (`PostHog::FeatureFlagsPoller#_request`, used by `_request_feature_flag_evaluation` for `POST /flags/?v=2`) has no retry. On any transient network error it rescues and immediately re-raises, surfacing a hard error to the caller.

By contrast, the event-capture path (`PostHog::Transport`) already retries with backoff (`retry_with_backoff`). The flags path is the inconsistent one.

### Why this matters

The flags request:
- opens a **fresh connection per request** (`Net::HTTP.start { ... }`, no keep-alive),
- has a **3s default timeout** (`feature_flag_request_timeout_seconds`),
- and **does not retry**.

So a single transient stall anywhere in the network path (packet loss + TCP retransmit, TLS setup jitter, an edge/proxy hiccup) that exceeds the 3s budget becomes a customer-visible `Net::ReadTimeout`, even when the PostHog flags service itself is healthy and responding in single-digit milliseconds.

We investigated a report of intermittent `Net::ReadTimeout` bursts against `POST /flags/?v=2`. Server-side the flags service was healthy throughout (fast 2xx responses, no deploy or pod churn during the window), and the gateway logged the customer's traffic as 2xx. The failures had no server-side trace, consistent with transient loss in the client → CDN → gateway path that a retry would have absorbed.

The error message the user sees comes from this path: `Unable to complete request to https://<host>/flags/?v=2`.

### Proposed fix

Add a bounded retry (e.g. 1 attempt with small backoff) to `_request` (or just the flag evaluation call) for idempotent transient errors that are already rescued there:

- `Net::ReadTimeout`
- `Net::OpenTimeout` / `Timeout::Error`
- `Errno::ECONNRESET`
- `EOFError`

Flag evaluation is side-effect-free, so retrying is safe. The existing `BackoffPolicy` used by `Transport` could be reused.

### Notes
- Still reproduces on `main` (no retry on `_request`). Recent versions did add `open_timeout`/`write_timeout`, but not retry.
- Affected user was on `3.5.4`; upgrading helps with the timeout settings but does not add retry.

### PR DESCRIPTION
**Pull request by haacked**
## :bulb: Motivation and Context

Fixes #195.

The feature flag request path (`FeatureFlagsPoller#_request`, used by `_request_feature_flag_evaluation` for `POST /flags/?v=2` and the definitions/remote-config reads) opened a fresh connection per request with a short timeout and no retry. On any transient network error it rescued and immediately re-raised, so a single blip in the client to CDN to gateway path (packet loss plus TCP retransmit, TLS setup jitter, an edge or proxy hiccup) that exceeded the timeout budget became a customer-visible `Net::ReadTimeout`, even when the flags service itself was healthy and responding in single-digit milliseconds.

The event-capture path (`Transport`) already retries with backoff, so the flags path was the inconsistent one.

This adds a bounded retry (default 1 retry) with backoff to `_request` for the transient, idempotent network errors it already rescues: `Timeout::Error` (includes `Net::OpenTimeout`), `Net::ReadTimeout`, `Net::WriteTimeout`, `Errno::ECONNRESET`, and `EOFError`. Flag evaluation is a side-effect-free read, so retrying is safe, and the existing `BackoffPolicy` is reused. The other rescued errors (`Errno::EINVAL`, `Net::HTTPBadResponse`, `Net::HTTPHeaderSyntaxError`, `Net::ProtocolError`) and connection-refused (`Errno::ECONNREFUSED`) still re-raise immediately, so upstream `$feature_flag_error` tracking is unchanged. After retries are exhausted the original exception still propagates. The retry count lives in `Defaults::FeatureFlags::FLAG_REQUEST_MAX_RETRIES`.

## :green_heart: How did you test it?

New `flags_spec` cases cover the three paths: retries once and succeeds after a transient `Net::ReadTimeout`, retries on `Errno::ECONNRESET` then re-raises once retries are exhausted, and does not retry on `Errno::ECONNREFUSED`. The existing timeout test in `feature_flag_spec` was updated to account for the extra retry request, with the result still `nil` so the intent is preserved. Full suite is green (553 examples) and rubocop is clean.

## :pencil: Checklist

- [x] I reviewed the submitted code.
- [x] I added tests to verify the changes.
- [ ] I updated the docs if needed.
- [ ] No breaking change or entry added to the changelog.

### If releasing new changes

- [x] Ran `pnpm changeset` to generate a changeset file

## 🤖 Agent context

**Autonomy:** Human-driven (agent-assisted)

Phil directed this fix from the GitHub issue. Implemented with Claude Code ([session](https://claude.ai/code/session_01Bhjg7hq4AVDWtTge5TWhdh)).

Decisions along the way: the retry is exception-only and deliberately does not copy `Transport`'s status-code retry, since `_request` returns the status code inside the parsed body and the upstream error-tracking layer turns 4xx/5xx into `api_error_*`. Adding status-code retry would have changed that behavior. The retryable set mirrors the transient subset already rescued and excludes `Errno::ECONNREFUSED` so a refused connection still fails fast. Placement is in `_request` rather than only the evaluation call because all three callers are idempotent reads.

### CODE REVIEW

File: spec/posthog/flags_spec.rb [hunk 1]
```diff
@@ -230,6 +230,43 @@ module PostHog
...
+          .to_raise(Errno::ECONNREFUSED)
+
+        expect { poller.get_flags('test-distinct-id') }.to raise_error(Errno::ECONNREFUSED)
+        expect(a_request(:post, flags_endpoint)).to have_been_made.times(1)
+      end
+    end
+
     it 'handles quota limited responses v3' do
       quota_limited_response = {
         flags: {},
```
[the remark this answers is not in the document]
Comment (haacked) [reply spec/posthog/flags_spec.rb:319]: Thanks for the suggestion, but the existing tests do cover the `<=` vs `<` boundary. With `FLAG_REQUEST_MAX_RETRIES = 1`, the condition `attempts <= 1` is `true` (allowing a retry) while `attempts < 1` would be `false` (blocking all retries) — these are distinct behaviors. The test `have_been_made.times(2)` would fail if the operator were incorrectly written as `<`. The test that expects exactly 2 requests directly verifies the boundary condition at the default retry limit.

File: lib/posthog/feature_flags.rb [hunk 2]
```diff
@@ -1279,20 +1295,24 @@ def _request(uri, request_object, timeout = nil, include_etag: false)
...
+        end
+        logger.debug("Unable to complete request to #{uri}")
+        raise
+      rescue Errno::EINVAL,
              Net::HTTPBadResponse,
              Net::HTTPHeaderSyntaxError,
-             Net::ReadTimeout,
-             Net::WriteTimeout,
              Net::ProtocolError
         logger.debug("Unable to complete request to #{uri}")
```
[the remark this answers is not in the document]
Comment (haacked) [reply lib/posthog/feature_flags.rb]: Good catch — fixed. The `rescue Errno::EINVAL / Net::HTTPBadResponse / Net::HTTPHeaderSyntaxError / Net::ProtocolError` block now also uses `_mask_tokens_in_url` to mask tokens before logging.

**Comment by marandaneto** [comment]
This retry happens synchronously in the feature flag request path, so callers of get_feature_flag is_feature_enabled will block during the backoff sleep and retry. That may be okay since it’s bounded to one retry, but it does increase worst-case latency under failures. Should we document this latency tradeoff or make the retry count configurable? (or even opt-in/out?)

the other point:

Per-call retry is simple and acceptable because it is bounded to one retry, but under a broad connectivity issue every request thread will retry independently. If we want stronger protection, a circuit breaker or retry throttle would be safer than making retry attempts/backoff global, since global mutable attempts can cause cross-request interference and thread-safety issues.

Example behavior:

 - Each call still gets its own normal attempt.
 - On transient failure, allow bounded retry.
 - Track consecutive transient failures globally.
 - After threshold, open circuit for e.g. 5–30s.
 - While circuit is open, fail fast / return default / local evaluation instead of sleeping and retrying.
 - After cooldown, allow one “half-open” probe.
 - If probe succeeds, close circuit and reset failures.

 That avoids this bad outage behavior:

 ```text
   100 incoming requests
   => 100 initial /flags calls
   => 100 sleeps
   => 100 retries
 ```

**Comment by marandaneto** [comment]
i have a few drafts open to port this fix but didnt test/review yet, will wait this one to land + resolve [this](https://github.com/PostHog/posthog-ruby/pull/198/#issuecomment-4796705507) comment if needed

 - posthog-js: https://github.com/PostHog/posthog-js/pull/3961
 - posthog-rs: https://github.com/PostHog/posthog-rs/pull/156
 - posthog-dotnet: https://github.com/PostHog/posthog-dotnet/pull/245
 - posthog-go: https://github.com/PostHog/posthog-go/pull/240
 - posthog-php: https://github.com/PostHog/posthog-php/pull/184
 - posthog-android: https://github.com/PostHog/posthog-android/pull/587
 - posthog-ios: https://github.com/PostHog/posthog-ios/pull/674
 - posthog-unity: https://github.com/PostHog/posthog-unity/pull/101

**Comment by haacked** [comment]
Yeah, a configurable retry makes sense.  I added a `feature_flag_request_max_retries` option (defaults to 1, set to 0 to opt out entirely) and documented the latency tradeoff on the option.

Regarding the circuit breaker, I agree that would offer stronger protection. But I think that should be a separate issue. Its more than I want to take on in this PR. 😄

**Review by marandaneto — APPROVED** [review]

**Review by turnipdabeets — APPROVED** [review]

**Review by turnipdabeets — APPROVED** [review]
